tokenizers/src/utils/regex.rs

```diff
@@ -0,0 +1,497 @@
+// All code and comments below this first line are copied from here (with some edits): https://github.com/bminixhofer/nlprule/blob/main/nlprule/src/utils/regex.rs
+
+
+
+//! An abstraction on top of Regular Expressions to add support for Serialization and
+//! to modularize the Regex backend.
+//! Adapts the approach from https://github.com/trishume/syntect/pull/270 with feature flags for the
+//! different backends.
+
+use once_cell::sync::OnceCell;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::hash::{Hash, Hasher};
+
+pub use regex_impl::{CaptureMatches, Captures, Match, Matches};
+
+#[derive(Debug)]
+pub struct Regex {
+    regex_str: String,
+    regex: OnceCell<regex_impl::Regex>,
+}
+
+impl Clone for Regex {
+    fn clone(&self) -> Self {
+        Regex {
+            regex_str: self.regex_str.clone(),
+            regex: OnceCell::new(),
+        }
+    }
+}
+
+impl Serialize for Regex {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(&self.regex_str)
+    }
+}
+
+impl<'de> Deserialize<'de> for Regex {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let regex_str = String::deserialize(deserializer)?;
+        Ok(Regex::new(regex_str))
+    }
+}
+
+impl Hash for Regex {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.regex_str.hash(state);
+    }
+}
+
+impl Regex {
+    /// Create a new regex from the pattern string.
+    ///
+    /// Note that the regex compilation happens on first use, which is why this method does not
```

Comment:
I really think this library wants to compile the regex on load, not on-use, that changes this to a Result.

It also probably enables your to pass `&str`  since the Regex compilation will create the object without making the copy and doesn't need to keep a reference to the string, (so no lifetime issues here it seems).